### ISSUE TITLE
Opening directories has a noticeable delay, UI is choppy in general

### ISSUE THREAD

**Issue by slikts** [fman-community/fman#220]
Windows 10, opening directories for the first time has a noticable delay, increased by the directories being on an HDD instead of SSD and by having a larger number of files, but even with a single file and an SSD there's a noticable flash of an empty list, and also a noticable delay before the current position in the list is highlighted. Scrolling and resizing are also choppy. I have a reasonably specced machine and everything else works fine, even much more complex software, so this suggests a lack of optimization.

One thing I'd suggest considering is [indexing files using NTFS MFT](https://en.wikipedia.org/wiki/Everything_(software)#How_it_works) to make lookups fast.

**Comment by mherrmann** [comment]
Performance can certainly be improved. Just for reference: What exactly are the specs of your machine?

**Comment by slikts** [comment]
i5 2500K@4.2, 16G RAM.

I find fman aggravatingly slow in the current state.

**Comment by witszymanski** [comment]
Hi, I started to test fman Yesterday and it's very slow on Windows 7 Professional. Getting into and out of the folder is way slower than any other file manager.

**Comment by ar-jan** [comment]
I posted a [profile](https://github.com/fman-users/fman/issues/362#issuecomment-430659734) in this related thread, maybe it's useful here too.

**Comment by witszymanski** [comment]
Just wanted to add that on my linux computer, the performance is fast. So it;s a Windows issue only probably.

**Comment by mherrmann** [comment]
Thank you @kaniawave. I'm happy it works well for you. My suspicion is that it is less dependent on the OS than on the hardware of the computer.

**Comment by crispinb** [comment]
> Thank you @kaniawave. I'm happy it works well for you. My suspicion is that it is less dependent on the OS than on the hardware of the computer.

One data point to suggest otherwise: I've been using fman on a fast laptop (i9, 32GB RAM, SSD) running Windows 10. I've recently added a Linux partition, and installed fman there - it's notably snappier. It starts faster, and directories with large numbers of files list instantly, which they didn't on Windows. 

This could in part be down to NTFS - I have my windows partitions mounted under linux, and when I use fman to browse to (say) c:\windows\system32, it's slower to list than most linux-native (ext4) directories. Still seems faster than under windows, but I can't be sure without a more objective test.

This doesn't appear to be unique to fman - other python-based GUI apps (eg Calibre) are much quicker on Linux too.

In any case, the overall effect is very clear: on my machine, fman is distinctly quicker under linux than windows.

**Comment by Raikiri** [comment]
I did a bunch more experiments to try and figure out some clues on why it's slow.
- It takes a lot of time to load folder contents for the first time. Around 2 seconds. However,
- Consequent attempts to load the folder take less time, around 0.5 second. And
- It takes the same 0.5 second to open that folder after restarting fman.
- It does not seem to be related to windows defender

This leads me to a conclusion that fman runs some kind of indexing when opening a folder for the very first time that seems to be laggy. However, even after everything is indexed, it still takes way more time to open a folder than in total commander, which seems to be almost instantenous. Just exiting a folder and then entering it again takes 0.5 second which's slower than usable unfortunately.

**Comment by mherrmann** [comment]
Yes, fman does some internal caching. I think the main problem currently is the rendering engine. With many files (like you found out), it's simply too slow.

**Comment by Raikiri** [comment]
> Yes, fman does some internal caching. I think the main problem currently is the rendering engine. With many files (like you found out), it's simply too slow.

I honestly don't think so. I tested on a bunch of folders with very modest amount of files: 10 images 1mb each, some sources, etc. And every time I opened a folder for the first time, it had a delay about 2000ms. When I opened the same folder again (even after closing fman and launching it again), it was noticeably faster (~500ms). Even empty folders have this delay.

It can be a completely separate issue to that rendering problem you're referring to, or interference of multiple factors. I can do some profiling (insturctions in this thread: https://github.com/fman-users/fman/issues/362 ) if that's needed and relevant.

**Comment by jspohr** [comment]
Hi @mherrmann , just wanted to check if there's any progress on this. I bought fman because I really liked the user interface. However, I stopped using it because of the sluggishness on Windows. On my 8 years old MacBook, fman runs circles around my Windows workstations which are much more powerful. Every interaction seems much slower than it should be:
- Startup: ~3s
- First tabbing between panes: ~1s
- Enumerating a directory with 10 files: ~1s
- With 50 files: ~5s
- There are frequent short pauses navigating the files list.
- EDIT: Copying a 1GB file from one SDD to another: 5.5s (2.5s in Windows Explorer)
- EDIT: Copying a directory structure with 1.200 files (86mb), 35 folders: ~15s (5s in Windows Explorer)

This happens on both of my Windows machines:
- Core i7 3.5GHz, 16GB RAM, SSD
- Xeon 3.6GHz, 64GB RAM, SSD

Seeing that nothing changed about that in a long time, I just wondered if I have to find an alternative. fman is a wonderful design, but the Windows implementation is a showstopper for me, unfortunately.
Thanks!

**Comment by mherrmann** [comment]
Hey @jspohr, I don't have any news I'm afraid. fman's current revenue level of a few hundred Euros per month only allows me to do the necessary maintenance to keep fman up-to-date on the various OSs, and to fix critical bugs.

**Comment by 5472qaywsx** [comment]
Hi @mherrmann, thank you for your software, but I have to disagree with this last comment of you: speed (or, on the contrary, delay and various retardation, lag times and the sort) are (at least, for me) very important to judge on the usability of this kind software (file manager); and if the usability is compromised, then I'm not keen on investing, even not the 18 € you ask for it (which is little, I agree).
So, you won't fix speed issues on MS-Windows (and there are many still open), but then how do you think to attract users/buyers?
Sorry if it's harsh, but that's my, hopefully clear, opinion.

**Comment by mherrmann** [comment]
I like clarity. The state is as before: fman's few hundred € per month are not enough for me to spend the significant time to work on this issue. If you don't like it for 18€, you don't have to buy or use it.

**Comment by 5472qaywsx** [comment]
Ok, thank you for your reply.
If you don't mind, I will continue to use it on an irregular basis, to evaluate it in different (personal) use cases. I also use other file manager like that, fundamentally because I think Windows Explorer is worth nothing.

**Comment by mherrmann** [comment]
Sure; I don't mind.

**Comment by granthusbands** [comment]
I think icons might be a huge part of this, at least on my system. As a potential simple solution, is it possible we could have a mode where fman provides its own simple icons rather than using the shell ones? I believe windows shell will happily read a lot of file data, even inside subfolders, when choosing icons.

Relatedly, this issue, and possibly #674, affect responsiveness to keypresses, so I have to keep slowing down to see how many keypresses actually got processed correctly. So rather than tapping backspace backspace down tab backspace backspace down (to move both panes a particular way), I have to hit backspace, backspace, wait for list, down, tab, wait for pane switch, backspace, backspace, wait for list, down. I think it's because keypresses sent to not-fully-initialized views get ignored.
